### PR TITLE
mgr/dashboard_v2: add health page and its dependencies

### PR DESCRIPTION
**Pull request by tspmelo**
Signed-off-by: Tiago Melo <tmelo@suse.com>

Below you can see the current health page in the dashboard_v2.
I would appreciate some suggestions regarding the colors, organization, etc...

dash health
localhost_4200_ fhd

**UPDATE**: new screenshot

### CODE REVIEW

**Comment by LenzGr** [comment]
Looks good already! At a first glance, I wonder if the "21 Objects" should be horizontally aligned with the two ring charts. Also, the "Health" widget is eating up a lot of real estate in case the cluster is in HEALTH_OK state. I like the colour scheme so far.

**Comment by jcsp** [comment]
I'm a bit surprised that doing one page involves touching 48 files -- I'm curious: is it mandatory to put so many functions in their own files?  Also, is there a special editor that people are using to keep track of all this?

**Comment by tspmelo** [comment]
Hi @jcsp ,
We try to follow the official angular styleguide.
As you can read on [1] they recommend that each file should have only 1 component/services/pipe.

This makes each file smaller/easier to maintain and each service/pipe can be easily reused.

I guess we could reduce that number of files if we refactor the pipes that are only used in 1 component into functions, but when I converted from the old dashboard to this one I decide to migrate all the old rivets “transformers” into “pipes” so we could continue to use the pipe in the html.

Another reason for the huge number of files is that when using Angular CLI[2] it will create automatically files for unit test. On this particular PR we have 14 spec files. 

In regards to the editor, I use VS Code and I know that most of us use WebStorm/PyCharm. Both of this editors have support for Angular intellisense.
For creating the files we use the Angular CLI [2]


[1] https://angular.io/guide/styleguide#single-responsibility
[2] https://github.com/angular/angular-cli

File: src/pybind/mgr/dashboard_v2/frontend/src/app/ceph/dashboard/mgr-summary.pipe.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
+import { Pipe, PipeTransform } from '@angular/core';
```
Comment (votdev) [inline src/pybind/mgr/dashboard_v2/frontend/src/app/ceph/dashboard/mgr-summary.pipe.ts:1]: I suggest to apply the following patch:
```
--- src/pybind/mgr/dashboard_v2/frontend/src/app/ceph/dashboard/mgr-summary.pipe.ts	(date 1516974115000)
+++ src/pybind/mgr/dashboard_v2/frontend/src/app/ceph/dashboard/mgr-summary.pipe.ts	(date 1517404213000)
@@ -1,4 +1,5 @@
 import { Pipe, PipeTransform } from '@angular/core';
+import * as _ from 'lodash';
 
 @Pipe({
   name: 'mgrSummary'
@@ -6,7 +7,7 @@
 export class MgrSummaryPipe implements PipeTransform {
   transform(value: any, args?: any): any {
     let result = '';
-    result += 'active: ' + value.active_name;
+    result += 'active: ' + (_.isUndefined(value.active_name) ? 'n/a' : value.active_name);
     if (value.standbys.length) {
       result += ', ' + value.standbys.length + ' standbys';
     }
```

**Review by LenzGr — DISMISSED** [review]
Looks good to me - I like the new layout.

**Review by votdev — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/dashboard_v2/frontend/src/app/core/navigation/navigation/navigation.component.html [hunk 2]
```diff
@@ -25,13 +25,17 @@
...
        id="bs-example-navbar-collapse-1">
     <ul class="nav navbar-nav navbar-primary">
-      <!-- <li routerLinkActive="active"
+      <li routerLinkActive="active"
           class="tc_menuitem tc_menuitem_dashboard">
         <a i18n
-           routerLink="/dashboard">Dashboard
+           routerLink="/dashboard">
+          <i class="fa fa-heartbeat fa-fw" [ngStyle]="topLevelData?.health_status | healthColor"></i>
+          <!--  -->
```
Comment (ricardoasmarques) [inline src/pybind/mgr/dashboard_v2/frontend/src/app/core/navigation/navigation/navigation.component.html]: This line should be deleted

File: src/pybind/mgr/dashboard_v2/frontend/src/app/shared/models/health.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
+export class Health {
```
Comment (ricardoasmarques) [inline src/pybind/mgr/dashboard_v2/frontend/src/app/shared/models/health.ts]: We already have one model file that uses the suffix `.model.ts`.
You should rename this file to `health.model.ts` or rename the existing file to `credentials.ts`, for consistency.

File: src/pybind/mgr/dashboard_v2/frontend/src/app/app-routing.module.ts [hunk 4]
```diff
@@ -1,17 +1,23 @@
...
 import { LoginComponent } from './core/auth/login/login.component';
 import { HostsComponent } from './ceph/cluster/hosts/hosts.component';
 
 const routes: Routes = [
-  { path: '', redirectTo: 'hosts', pathMatch: 'full' },
+  { path: '', redirectTo: 'dashboard', pathMatch: 'full' },
+  {
+    path: 'dashboard',
+    canActivate: [AuthGuardService],
+    component: DashboardComponent
```
Comment (ricardoasmarques) [inline src/pybind/mgr/dashboard_v2/frontend/src/app/app-routing.module.ts]: Can you switch line 12 and 13 to use the same order that we use on `login` and `hosts` (1 path, 2 component, 3 canActivate)?

File: src/pybind/mgr/dashboard_v2/frontend/src/app/ceph/dashboard/health/health.component.spec.ts [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+  const dashboardServiceStub = {
+    getHealth() {
+      return {};
+    }
+  };
+  beforeEach(
+    async(() => {
+      TestBed.configureTestingModule({
+        declarations: [
+          // HealthComponent,
```
Comment (ricardoasmarques) [inline src/pybind/mgr/dashboard_v2/frontend/src/app/ceph/dashboard/health/health.component.spec.ts]: Remove commented code.

File: src/pybind/mgr/dashboard_v2/frontend/src/app/ceph/dashboard/mgr-summary.pipe.ts (hunk 1 above)
Comment (tspmelo) [reply src/pybind/mgr/dashboard_v2/frontend/src/app/ceph/dashboard/mgr-summary.pipe.ts:1]: done

File: src/pybind/mgr/dashboard_v2/controllers/dashboard.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+            {
+                "id": f['id'],
+                "name": f['mdsmap']['fs_name']
+            }
+            for f in fsmap['filesystems']
+        ]
+
+        return {
+            'health_status': self.health_data()['status'],
+            'filesystems': filesystems,
```
Comment (sebastian-philipp) [inline src/pybind/mgr/dashboard_v2/controllers/dashboard.py:72]: Hm. exporting `filesystems` in `/api/dashboard/toplevel` seems strange to me.

File: src/pybind/mgr/dashboard_v2/controllers/dashboard.py [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+        return osd_map
+
+    def health_data(self):
+        health_data = self.mgr.get("health")
+        health =  json.loads(health_data['json'])
+
+        # Transform the `checks` dict into a list for the convenience
+        # of rendering from javascript.
+        checks = []
+        for k, v in health['checks'].iteritems():
```
Comment (sebastian-philipp) [inline src/pybind/mgr/dashboard_v2/controllers/dashboard.py]: Please change to: `for k, v in health['checks'].items():` (Python 3 compatibility)

File: src/pybind/mgr/dashboard_v2/controllers/dashboard.py [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+            v['type'] = k
+            checks.append(v)
+
+        checks = sorted(checks, cmp=lambda a,
+                        b: a['severity'] > b['severity'])
```
Comment (sebastian-philipp) [inline src/pybind/mgr/dashboard_v2/controllers/dashboard.py]: `cmp` parameter was removed in Python 3. Use `sorted(checks, lambda c:c['severity'])` instead.

File: src/pybind/mgr/dashboard_v2/frontend/src/app/shared/models/health.ts (hunk 3 above)
Comment (tspmelo) [reply src/pybind/mgr/dashboard_v2/frontend/src/app/shared/models/health.ts]: I change credentials file to credentials.ts.
Angular uses that naming for their models [1].

[1] https://angular.io/tutorial/toh-pt1#create-a-hero-class

File: src/pybind/mgr/dashboard_v2/frontend/src/app/app-routing.module.ts (hunk 4 above)
Comment (tspmelo) [reply src/pybind/mgr/dashboard_v2/frontend/src/app/app-routing.module.ts]: done

File: src/pybind/mgr/dashboard_v2/frontend/src/app/ceph/dashboard/health/health.component.spec.ts (hunk 5 above)
Comment (tspmelo) [reply src/pybind/mgr/dashboard_v2/frontend/src/app/ceph/dashboard/health/health.component.spec.ts]: done

File: src/pybind/mgr/dashboard_v2/controllers/dashboard.py (hunk 6 above)
Comment (tspmelo) [reply src/pybind/mgr/dashboard_v2/controllers/dashboard.py:72]: I will be using that information on the filesystem menu.

File: src/pybind/mgr/dashboard_v2/controllers/dashboard.py (hunk 7 above)
Comment (tspmelo) [reply src/pybind/mgr/dashboard_v2/controllers/dashboard.py]: done

File: src/pybind/mgr/dashboard_v2/controllers/dashboard.py (hunk 8 above)
Comment (tspmelo) [reply src/pybind/mgr/dashboard_v2/controllers/dashboard.py]: done

**Comment by tspmelo** [comment]
Added translation attributes to the health page.

**Review by ricardoasmarques — DISMISSED** [review]

**Review by jecluis — COMMENTED** [review]
@tspmelo Is the new code you're adding to the backend being exercised by any of our tests? If not, mind adding corresponding tests?

**Review by votdev — DISMISSED** [review]

**Comment by jecluis** [comment]
@tspmelo what is holding this PR? It seems there are two other PRs depending on this one.

**Comment by tspmelo** [comment]
Just pushed a commit that introduces unit tests for the backend and fixes a frontend problem, the toplevel service was trying to request data even when user lost its session.

**Review by jecluis — APPROVED** [review]

**Pull request merged**